src/tests/daos_mw_fi.c

```diff
@@ -0,0 +1,264 @@
+
+		entry = daos_prop_entry_get(prop, DAOS_PROP_CO_ROOTS);
+		D_ASSERT(entry != NULL);
+		roots = (struct daos_prop_co_roots *)entry->dpe_val_ptr;
+		if (daos_obj_id_is_nil(roots->cr_oids[0]) ||
+		    daos_obj_id_is_nil(roots->cr_oids[1])) {
+			fprintf(stderr, "Failed: Invalid superblock or root object ID\n");
+			return -DER_INVAL;
+		}
+		return punch_obj(coh, roots->cr_oids[0], NULL);
+	}
+
+	if (path[0] != '/') {
+		fprintf(stderr, "Failed: Path must be absolute from the container root\n");
+		return -DER_INVAL;
+	}
+
+	D_STRNDUP(dir, path, PATH_MAX);
+	if (dir == NULL)
+		return -DER_NOMEM;
+	D_STRNDUP(file, path, PATH_MAX);
+	if (file == NULL)
+		D_GOTO(out_dfs, rc = -DER_NOMEM);
+
+	rc = dfs_mount(poh, coh, O_RDWR, &dfs);
+	if (rc) {
+		fprintf(stderr, "dfs_mount() failed: (%d)\n", rc);
+		D_GOTO(out_path, rc);
+	}
+
+	fname = basename(file);
+	dirp = dirname(dir);
+	printf("punching %s from %s\n", fname, dirp);
+	rc = dfs_lookup(dfs, dirp, O_RDWR, &obj, NULL, NULL);
+	if (rc) {
+		fprintf(stderr, "dfs_lookup() failed: (%d)\n", rc);
+		D_GOTO(out_dfs, rc);
+	}
+	rc = dfs_obj2id(obj, &oid);
+	dfs_release(obj);
+	if (rc) {
+		fprintf(stderr, "dfs_obj2id() failed: (%d)\n", rc);
+		D_GOTO(out_dfs, rc);
+	}
+
+	rc = punch_obj(coh, oid, fname);
+out_dfs:
+	dfs_umount(dfs);
+out_path:
+	D_FREE(file);
+	D_FREE(dir);
+	return rc;
+}
+
+static int
+fi_pydaos(daos_handle_t poh, daos_handle_t coh, const char *name, daos_prop_t *prop)
+{
+	daos_handle_t			oh;
+	struct daos_prop_entry		*entry;
+	struct daos_prop_co_roots	*roots;
+	daos_key_t			dkey;
+	int				rc;
+
+	entry = daos_prop_entry_get(prop, DAOS_PROP_CO_ROOTS);
+	D_ASSERT(entry != NULL);
+	roots = (struct daos_prop_co_roots *)entry->dpe_val_ptr;
+	if (daos_obj_id_is_nil(roots->cr_oids[0])) {
+		fprintf(stderr, "Failed: Invalid PyDAOS root object ID\n");
+		return -DER_INVAL;
+	}
+	roots->cr_oids[0].hi |= (uint64_t)DAOS_OT_KV_HASHED << OID_FMT_TYPE_SHIFT;
+
+	/** punch the dkey entry leaving the object it corresponds to. */
+	rc = daos_obj_open(coh, roots->cr_oids[0], DAOS_OO_RW, &oh, NULL);
+	if (rc) {
+		fprintf(stderr, "daos_obj_open() failed: "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	d_iov_set(&dkey, (void *)name, strlen(name));
+
+	rc = daos_obj_punch_dkeys(oh, DAOS_TX_NONE, DAOS_COND_PUNCH, 1, &dkey, NULL);
+	if (rc) {
+		daos_obj_close(oh, NULL);
+		fprintf(stderr, "daos_obj_punch_dkeys() failed: "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+	rc = daos_obj_close(oh, NULL);
+	if (rc)
+		fprintf(stderr, "daos_obj_close() failed: "DF_RC"\n", DP_RC(rc));
+	return rc;
+}
+
+int
+main(int argc, char **argv)
+{
+	daos_handle_t			poh, coh;
+	uint32_t			props[] = {DAOS_PROP_CO_LAYOUT_TYPE, DAOS_PROP_CO_ROOTS};
+	const int			num_props = ARRAY_SIZE(props);
+	int				i;
+	daos_prop_t			*prop = NULL;
+	struct daos_prop_entry		*entry = NULL;
+	int				rc;
+
+	if (argc != 5 && argc != 4) {
+		fprintf(stderr,
+			"usage: ./daos_mw_fi pool cont [punch_entry; punch_sb] [name or path]\n");
+		exit(1);
```

Comment:
Curious - do we prefer `exit` in `main` over `return`? (basically the same)

Reply:
i have no preference